File: solutions/kotlin/knapsack/1/src/main/kotlin/Knapsack.kt
data class Item(val weight: Int, val value: Int)

fun knapsack(maximumWeight: Int, items: List<Item>): Int {
    var d: IntArray = IntArray(maximumWeight + 1) {
        if (it == 0) {
            0
        } else {
            -1
        }
    };
    for (item in items) {
        for (weight in 0..(maximumWeight - item.weight)) {
            if (d[maximumWeight - item.weight - weight] != -1) {
                d[maximumWeight - weight] =
                    if (d[maximumWeight - weight] >= d[maximumWeight - item.weight - weight] + item.value) {
                        d[maximumWeight - weight]
                    } else {
                        d[maximumWeight - item.weight - weight] + item.value
                    }
            }
        }
    }
    var ans = -1;
    for (i in d) {
        if (i > ans) {
            ans = i
        }
    }
    return ans
}

File: solutions/ocaml/knapsack/1/knapsack.ml
type item = {
    weight : int;
    value : int;
}

let rec update (items: item list) (result: int array) =
  match items with
  | [] -> ()
  | i::is ->
    let () = update is result in
    for j = Array.length result - 1 downto 0  do
      if j + i.weight <= Array.length result - 1 && result.(j) <> -1 then
        result.(j+i.weight) <- max result.(j+i.weight) (result.(j) + i.value)
    done


let maximum_value (items: item list) (capacity: int): int =
  let result = Array.init (capacity+1) (fun i -> if i = 0 then 0 else -1) in
  let () = update items result in
  let ans = ref 0 in
  let () = for i = 0 to capacity do
    if !ans < result.(i) then
      ans := result.(i)
  done
  in !ans
